multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+    {
+        data_stream << *destination;
+    }
+
+    for (const auto& source : m_sources)
+    {
+        data_stream << *source->script_witness;
+    }
+
+    data_stream << **lock_time;
+
+    return make_clone(data_stream.get_content());
+}
+
+BitcoinNetType BitcoinTransactionSegWit::get_net_type() const
+{
+    return static_cast<BitcoinNetType>(get_blockchain_type().net_type);
+}
+
+void BitcoinTransactionSegWit::sign()
+{
+
+    for (auto& source : m_sources)
+    {
+        source->script_signature.reset();
+    }
+
+    BitcoinDataStream preview_inputs;
+    for (const auto& source : m_sources)
+    {
+        preview_inputs << *source->preview_transaction();
+    }
+    BitcoinDataStream preview_outputs;
+    for (const auto& destination : m_destinations)
+    {
+        preview_outputs << *destination;
+    }
+
+    BitcoinDataStream data;
+    for (const auto& source : m_sources)
+    {
+        data << **source->sequence;
+    }
+
+
+    hash<256> hash_prevouts = do_hash<SHA2_DOUBLE, 256>(preview_inputs.get_content());
+    hash<256> hash_sequence = do_hash<SHA2_DOUBLE, 256>(data.get_content());
+    hash<256> hash_outpusts = do_hash<SHA2_DOUBLE, 256>(preview_outputs.get_content());
+
+    for (auto& source : m_sources)
+    {
+        source->sign(hash_prevouts, hash_sequence, hash_outpusts, *lock_time);
+    }
+
+
+}
+
+void BitcoinTransactionSegWit::verify()
+{
+    if (m_sources.empty())
+    {
+        THROW_EXCEPTION2(ERROR_TRANSACTION_NO_SOURCES,
+                          "Transaction should have at least one source.");
+    }
+    if (m_destinations.empty())
+    {
+        THROW_EXCEPTION2(ERROR_TRANSACTION_NO_DESTINATIONS,
+                "Transaction should have at least one destination.");
+    }
+    std::string missing_properties;
+    if (!validate_all_properties(&missing_properties))
+    {
+        THROW_EXCEPTION2(ERROR_NOT_ALL_REQUIRED_PROPERTIES_SET,
+                "Not all required properties set.")
+                << "\n" << missing_properties << ".";
+    }
+
+    if (calculate_diff() < 0)
+    {
+        THROW_EXCEPTION2(ERROR_TRANSACTION_INSUFFICIENT_FUNDS,
+                "Transaction is trying to spend more than available in inputs")
+                << " : available - spent = "<< calculate_diff().get_value();
+    }
+
+    size_t source_index = 0;
+    for (const auto& s: m_sources)
+    {
+        const auto& public_key = s->private_key->make_public_key();
+        BinaryData public_key_hash_data;
+        std::array<uint8_t, HASH160_LEN> public_key_hash;
+        public_key_hash_data = as_binary_data(public_key_hash);
+        unsigned char segwit_script[HASH160_LEN+2] = {'\0'};
+
+        BinaryData hash_pub_key = power_slice(as_binary_data(segwit_script), 2, -2);
+
+        bitcoin_hash_160(public_key->get_content(), &hash_pub_key);
+        segwit_script[0] = 0x00;
+        segwit_script[1] = HASH160_LEN;
+        bitcoin_hash_160(as_binary_data(segwit_script), &public_key_hash_data);
```

Comment:
that looks strangely similar to the `make_script_pubkey`, but yet it is somewhat different. Please move common code to some function.